crates/core/tedge_mapper/src/c8y/converter.rs

```diff
@@ -224,28 +226,43 @@ where
         &mut self,
         input: &Message,
     ) -> Result<Vec<Message>, ConversionError> {
-        let tedge_event = ThinEdgeEvent::try_from(input.topic.name.as_str(), input.payload_str()?)?;
-        let c8y_event = C8yCreateEvent::try_from(tedge_event)?;
+        let mut vec = Vec::new();
+        let c8y_event;
+
+        let tedge_event = ThinEdgeEvent::try_from(&input.topic.name, input.payload_str()?)?;
+        let child_id = tedge_event.source.clone();
+
+        self.create_external_source_if_does_not_exist(&tedge_event, &mut vec);
+
+        c8y_event = C8yCreateEvent::try_from(tedge_event)?;
 
         // If the message doesn't contain any fields other than `text` and `time`, convert to SmartREST
         let message = if c8y_event.extras.is_empty() {
             let smartrest_event = Self::serialize_to_smartrest(&c8y_event)?;
             let smartrest_topic = Topic::new_unchecked(SMARTREST_PUBLISH_TOPIC);
-
             Message::new(&smartrest_topic, smartrest_event)
         } else {
             // If the message contains extra fields other than `text` and `time`, convert to Cumulocity JSON
             let cumulocity_event_json = serde_json::to_string(&c8y_event)?;
             let json_mqtt_topic = Topic::new_unchecked(C8Y_JSON_MQTT_EVENTS_TOPIC);
-
             Message::new(&json_mqtt_topic, cumulocity_event_json)
         };
 
+        // vec.len() == 1, means the child device is not registered with c8y cloud.
+        // So, before forwarding the message, check if the size is bigger than
+        // the supported MQTT threshold. If so, then ask to register the device first.
+        if vec.len() == 1 && message.payload_bytes().len() > self.size_threshold.0 {
```

Comment:
This is too fragile and obscure. Hence, I would look for another approach.

* The long comment is only required because the code is obscure.
* This is too fragile. Future extensions might break the assumption that `vec.len() == 1` implies `not registered`.
* Another source of fragility and obscurity: the check `message.payload_bytes().len() > self.size_threshold.0` is done in two places.

Why not, as we discussed, adding a `need_registration` boolean that makes things clearer?

```
let need_registration = self.create_external_source_if_does_not_exist(&tedge_event, &mut vec);

if message.payload_bytes().len() < self.size_threshold.0 {
   // The message can be sent via MQTT
  vec.push(message);
} else if !need_registration  {
  // The message must be sent over HTTP
  let _ = self.http_proxy.send_event(c8y_event).await?;
} else {
  // The message should be sent over HTTP but this cannot be done
 return Err(ConversionError::ChildDeviceNotRegistered {
                id: child_id.unwrap_or_else(|| "".into()),
            });
}
Ok(vec)
```

Furthermore, naming could help a lot to clarify the code.
* `messages` would be better than `vec`.
* `add_registration_messages` would be better than `create_external_source_if_does_not_exist`.
* `self.can_send_over_mqtt(&message)` would be better than `message.payload_bytes().len() < self.size_threshold.0`.

Reply:
Addressed all of them